clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -752,6 +758,106 @@ private StoragePartitionConfig getStoragePartitionConfig(
+          && Arrays.equals(consumerRecord.getKey().getKey(), KafkaKey.HEART_BEAT.getKey())) {
+        boolean isLeaderCompleteHeaderFound = false;
+        LeaderCompleteState oldState = partitionConsumptionState.getLeaderCompleteState();
+        LeaderCompleteState newState = oldState;
+        PubSubMessageHeaders pubSubMessageHeaders = consumerRecord.getPubSubMessageHeaders();
+        for (PubSubMessageHeader header: pubSubMessageHeaders.toList()) {
+          if (header.key().equals(VENICE_LEADER_COMPLETION_STATE_HEADER)) {
+            newState = LeaderCompleteState.valueOf(header.value()[0]);
+            partitionConsumptionState
+                .setLastLeaderCompleteStateUpdateInMs(consumerRecord.getValue().producerMetadata.messageTimestamp);
+            isLeaderCompleteHeaderFound = true;
+            break; // only interested in this header here
+          }
+        }
+        if (!isLeaderCompleteHeaderFound) {
+          // reset LeaderCompleteState: If a leader originally sent this header but later is rolled back to a version
+          // that doesn't support this header or this TP gets a new leader which doesn't support this header yet.
+          newState = LEADER_COMPLETE_STATE_UNKNOWN;
+        }
+
+        if (oldState != newState) {
+          LOGGER.info(
+              "LeaderCompleteState for store {} version {} partition {} changed from {} to {}",
+              storeName,
+              versionNumber,
+              partitionConsumptionState.getPartition(),
+              oldState,
+              newState);
+          partitionConsumptionState.setLeaderCompleteState(newState);
+        } else {
+          LOGGER.debug(
+              "LeaderCompleteState for store {} version {} partition {} received from leader: {} and is unchanged from the previous state",
+              storeName,
+              versionNumber,
+              partitionConsumptionState.getPartition(),
+              newState);
+        }
+        if (!partitionConsumptionState.isFirstHeartBeatSOSReceived()) {
+          partitionConsumptionState.setFirstHeartBeatSOSReceived(true);
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks whether the lag is acceptable for hybrid stores
+   *
+   * If the instance is a standby or DaVinciClient: Also check if
+   * 1. leaderCompleteStatus header has been received and
+   * 2. leader was completed and
+   * 3. the last update time was within 5 seconds
+   */
+  protected boolean isLagAcceptableForHybridStore(PartitionConsumptionState pcs, long offsetLag, long offsetThreshold) {
+    boolean isLagAcceptable = (offsetLag <= offsetThreshold);
+
+    if (isHybridMode() && (isDaVinciClient || pcs.getLeaderFollowerState().equals(STANDBY))) {
+      if (!isLagAcceptable) {
+        return false;
+      }
+
+      // non AA stores have issues reading HB SOS from the RT leading to the standby replicas waiting for
+      // leader completion state header indefinitely, so disabling it until that issue is resolved
```

Comment:
That's because leaders in some colos might be running on old codes? If so, we will have the same problem when doing canary deployment in one colo though.